binary-option/program/src/processor.rs

```diff
@@ -0,0 +1,743 @@
+            program_id.as_ref(),
+        ],
+        program_id,
+    );
+    let seeds = &[
+        long_token_mint_info.key.as_ref(),
+        short_token_mint_info.key.as_ref(),
+        token_program_info.key.as_ref(),
+        program_id.as_ref(),
+        &[bump_seed],
+    ];
+
+    if !binary_option.settled {
+        return Err(BinaryOptionError::BetNotSettled.into());
+    }
+    assert_owned_by(long_token_mint_info, &spl_token::id())?;
+    assert_owned_by(short_token_mint_info, &spl_token::id())?;
+    assert_keys_equal(collector_long_token_account.owner, *collector_info.key)?;
+    assert_keys_equal(collector_short_token_account.owner, *collector_info.key)?;
+    assert_keys_equal(collector_account.owner, *collector_info.key)?;
+    assert_keys_equal(escrow_owner_key, *escrow_authority_info.key)?;
+    assert_keys_equal(
+        *long_token_mint_info.key,
+        binary_option.long_mint_account_pubkey,
+    )?;
+    assert_keys_equal(
+        *short_token_mint_info.key,
+        binary_option.short_mint_account_pubkey,
+    )?;
+    assert_keys_equal(*escrow_account_info.key, binary_option.escrow_account_pubkey)?;
+    assert_keys_equal(*escrow_account_info.key, binary_option.escrow_account_pubkey)?;
+    assert_keys_equal(
+        collector_long_token_account.mint,
+        binary_option.long_mint_account_pubkey,
+    )?;
+    assert_keys_equal(
+        collector_short_token_account.mint,
+        binary_option.short_mint_account_pubkey,
+    )?;
+    assert_keys_equal(
+        collector_account.mint,
+        binary_option.escrow_mint_account_pubkey,
+    )?;
+
+    let winner_long = collector_long_token_account.mint == binary_option.winning_side_pubkey;
+    let winner_short = collector_short_token_account.mint == binary_option.winning_side_pubkey;
+    let reward = match [winner_long, winner_short] {
+        [true, false] => collector_long_token_account.amount,
+        [false, true] => collector_short_token_account.amount,
+        _ => return Err(BinaryOptionError::TokenNotFoundInPool.into()),
+    };
+
+    topup(
+        escrow_authority_info,
+        rent_info,
+        system_account_info,
+        fee_payer_info,
+        1,
+    )?;
+    spl_burn_signed(
+        &token_program_info,
+        &collector_long_token_account_info,
+        &long_token_mint_info,
+        &escrow_authority_info,
+        collector_long_token_account.amount,
+        seeds,
+    )?;
+    spl_burn_signed(
+        &token_program_info,
+        &collector_short_token_account_info,
+        &short_token_mint_info,
+        &escrow_authority_info,
+        collector_short_token_account.amount,
+        seeds,
+    )?;
+    if reward > 0 {
+        spl_token_transfer_signed(
+            &token_program_info,
+            &escrow_account_info,
+            &collector_account_info,
+            &escrow_authority_info,
+            reward * escrow_account.amount,
+            binary_option.circulation,
```

Comment:
How does `circulation` relate to `denom` as it's being used here?

Reply:
suppose my pool has 100 contracts (circulation = 100) that each settle to 100. The total size of the pool is 10k (escrow_account.amount = 10000). I collect for 20 contracts (reward = 20)

This means that my reward is 20 * 10000 / 100 = 2000

I then decrement the circulation by 20 so my end variable state is
escrow_account.amount = 8000
circulation = 80